core/arch/arm/mm/core_mmu_lpae.c

```diff
@@ -433,11 +500,53 @@ static uint64_t mattr_to_desc(unsigned level, uint32_t attr)
 	return desc;
 }
 
+static uint64_t *get_base_table(struct mmu_partition *prtn, size_t tbl_idx,
+				size_t core_pos)
+{
+	assert(tbl_idx < NUM_BASE_TABLES);
+	assert(core_pos < CFG_TEE_CORE_NB_CORE);
+
+	return  prtn->base_tables + (core_pos * NUM_BASE_TABLES + tbl_idx) *
+				    NUM_BASE_LEVEL_ENTRIES;
+}
+
+static uint64_t *get_l2_ta_tables(struct mmu_partition *prtn, size_t thread_id)
+{
+	assert(thread_id < CFG_NUM_THREADS);
+
+	return prtn->l2_ta_tables + XLAT_TABLE_ENTRIES * thread_id;
+}
+
+#if (CORE_MMU_BASE_TABLE_LEVEL == 0)
+static uint64_t *get_l1_ta_table(struct mmu_partition *prtn, size_t base_idx,
+				 size_t core_pos)
+{
+	size_t idx = 0;
+	uint64_t *tbl = NULL;
+
+	idx = prtn->user_l1_table_idx[core_pos * NUM_BASE_TABLES + base_idx];
+	tbl = (void *)((vaddr_t)prtn->xlat_tables + idx * XLAT_TABLE_SIZE);
+	return tbl;
+}
+
+static void set_l1_ta_table(struct mmu_partition *prtn, size_t base_idx,
+			    size_t core_pos, uint64_t *tbl)
+{
+	size_t idx = 0;
+
+	idx = ((vaddr_t)tbl - (vaddr_t)prtn->xlat_tables) / XLAT_TABLE_SIZE;
+	if (idx > MAX_XLAT_TABLES)
+		panic();
```

Comment:
The previous implementation tested `idx` index against the number of tables already consumed (`prtn->xlat_table_used`).  It is intentional to allow up to `MAX_XLAT_TABLES` now?

Reply:
Good catch, I'll restore the assert().

I'm having my doubts about this array of indices. With a hardcoded limit, it's easy to ensure that the index type is wide enough, but with a more dynamic approach, we may be better off with pointers. It's also slightly less complicated and that should be welcome in this file. If you have a VA space larger than 39 bits, you can afford 6-7 extra bytes per CPU (in addition to the extra 4k page needed). The decision can wait until we have a dynamic configuration of the number of cores.